src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java

```diff
@@ -309,6 +330,249 @@ public QuizExerciseResource(QuizExerciseRepository quizExerciseRepository, Parti
         return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(ENTITY_NAME, id.toString())).build();
     }
 
+    /**
+     * PUT  /quiz-exercises-re-evaluate : Re-evaluates an existing quizExercise.
+     *
+     * 1. reset not allowed changes and set flag updateResultsAndStatistics if a recalculation of results and statistics is necessary
+     * 2. save changed quizExercise
+     * 3. if flag is set: -> change results if an answer or a question is set invalid
+     *                    -> recalculate statistics and results and save them.
+     *
+     * @param quizExercise the quizExercise to re-evaluate
+     * @return the ResponseEntity with status 200 (OK) and with body the re-evaluated quizExercise,
+     * or with status 400 (Bad Request) if the quizExercise is not valid,
+     * or with status 500 (Internal Server Error) if the quizExercise couldn't be re-evaluated
+     * @throws URISyntaxException if the Location URI syntax is incorrect
+     */
+    @PutMapping("/quiz-exercises-re-evaluate")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    @Timed
+    public ResponseEntity<QuizExercise> reEvaluateQuizExercise(@RequestBody QuizExercise quizExercise) throws URISyntaxException {
+        log.debug("REST request to re-evaluate QuizExercise : {}", quizExercise);
+        if (quizExercise.getId() == null) {
+            return createQuizExercise(quizExercise);
+        }
+        QuizExercise originalQuizExercise = quizExerciseRepository.findOne(quizExercise.getId());
+        boolean updateOfResultsAndStatisticsNecessary = undoUnallowedChangesAndCheckIfRecalculationIsNecessary(quizExercise, originalQuizExercise);
+
+        //change existing results if an answer or and question was deleted
```

Comment:
The code below is still too complex You have again nested for loops with 4 levels. Please avoid this and refactor your code by
1) extracting methods into the right place (e.g. into the domain classes)
2) avoiding code duplication by writing more and smaller methods with a specific purpose